doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1257,409 +1261,343 @@ Managing the local context
+   .. tacn:: eenough ( @ident : @type ) {? by @ltac_expr3 }
+             eenough @one_type {? @as_ipat } {? by @ltac_expr3 }
+      :name: eenough; _
 
-.. tacv:: enough @type
+      Unlike :tacn:`enough`, the :n:`@type` and :n:`@one_type` in
+      :tacn:`eenough` may contain :term:`holes <hole>`, denoted by :n:`_`,
+      for which the tactic will create existential variables.  This lets you
+      avoid specifying the asserted statement completely until you start to use
+      the hypothesis or later start to prove the statement.
 
-   This behaves like :n:`enough (@ident : @type)` with the name :token:`ident` of
-   the hypothesis generated by Coq.
+.. tacn:: cut @one_type
 
-.. tacv:: enough @type as @simple_intropattern
+   Implements the non-dependent case of the :ref:`App <app_rule>` typing rule,
+   the Modus Ponens inference rule.  It is equivalent to
+   :n:`enough (@ident: @one_type). revert @ident.`
+   This tactic is generally considered obsolete but it is still widely
+   used in old scripts.
 
-   This behaves like :n:`enough @type` using :token:`simple_intropattern` to name or
-   destruct the new hypothesis.
+.. tacn:: pose proof @term {? @as_ipat }
+          pose proof ( @ident := @term )
+   :name: pose proof; _
 
-.. tacv:: enough (@ident : @type) by @tactic
-          enough @type {? as @simple_intropattern } by @tactic
+   The first form behaves like :n:`assert @one_type {? @as_ipat } by exact @term`
+   where :token:`one_type` is the type of :token:`term`.
 
-   This behaves as above but with :token:`tactic` expected to solve the initial goal
-   after the extra assumption :token:`type` is added and possibly destructed. If the
-   :n:`as @simple_intropattern` clause generates more than one subgoal, :token:`tactic` is
-   applied to all of them.
+   .. Théo notes it's odd that the first form uses @term instead of @one_term
 
-.. tacv:: eenough @type {? as @simple_intropattern } {? by @tactic }
-          eenough (@ident : @type) {? by @tactic }
-   :name: eenough; _
+   The second form is equivalent to :n:`assert (@ident := @term)`.
 
-   While the different variants of :tacn:`enough` expect that no existential
-   variables are generated by the tactic, :tacn:`eenough` removes this constraint.
+   .. tacn:: epose proof @term {? @as_ipat }
+             epose proof ( @ident := @term )
+      :name: epose proof; _
 
-.. tacv:: cut @type
-   :name: cut
+      While :tacn:`pose proof` expects that no existential variables are generated by
+      the tactic, :tacn:`epose proof` removes this constraint.
 
-   This tactic applies to any goal. It implements the non-dependent case of
-   the “App” rule given in :ref:`typing-rules`. (This is Modus Ponens inference
-   rule.) :n:`cut U` transforms the current goal :g:`T` into the two following
-   subgoals: :g:`U -> T` and :g:`U`. The subgoal :g:`U -> T` comes first in the
-   list of remaining subgoal to prove.
+.. tacn:: specialize @one_term_with_bindings {? @as_ipat }
```

Comment:
Let's come up with examples before I try to reword the description.

Example 1:
```
Goal forall x: nat, (forall n, n >= 0) -> True.
intros.  (* H is forall n : nat, n >= 0 *)
specialize (H x). (* H is x >= 0 *)
```

2. In `List.v`, `remove_length_lt` (line 764) add ". " after the `specialize`, can you explain how the `_` in `specialize (IHl _ Hin)` is used, what it means?

![image](https://user-images.githubusercontent.com/1253341/195702708-66d67ad1-bcc3-4af1-aa04-3470f652bc14.png)

Are there other interesting examples?

Reply:
> 2\. In `List.v`, `remove_length_lt` (line 764) add ". " after the `specialize`, can you explain how the `_` in `specialize (IHl _ Hin)` is used, what it means?

This is simply a case of a hole being left by the user because it can be inferred by Coq. (The variable `x0` appears in the premise, so by providing `Hin`, Coq is able to infer the value of `x0`.) This is also a common pattern when passing arguments to a lemma in a call to `apply`, etc. Nothing particularly specific to `specialize` here.

> Are there other interesting examples?

I would say that this is representative of the most interesting use case for `specialize` but there might be some edge cases that are not covered. One such edge case is when you call `specialize` on something that is a lemma and not a hypothesis from the local context. In this case, if you provide the `as` clause you can name (or further transform) the result and add it to the local context, but if you don't, then the "specialized" lemma is added as a premise to the goal:

```coq
Require Import Arith.

Goal True.
specialize (Nat.add_comm 1 0).
(* Goal is now: 1 + 0 = 0 + 1 -> True *)
```

Reply:
How do you know when `specialize` will modify the goal (as in your example immediately above) versus modifying a hypothesis?

And it appears that what you see afterwards is the type of the provided term, not that term itself.

Reply:
> How do you know when `specialize` will modify the goal (as in your example immediately above) versus modifying a hypothesis?

It modifies the goal (adds a premise to the goal) only if the head of the term passed as an argument is not the name of a hypothesis from the local context *and* there is no `as` clause.

> And it appears that what you see afterwards is the type of the provided term, not that term itself.

That's correct. `specialize (some_lemma args) as H` is the same thing as `pose proof (some_lemma args) as H`.

Reply:
I think it's time to create a section that starts to explain the relationship between terms and their types, e.g. if the type of H is `forall n m : nat, n >= m` then `(H x)` is like a partial evaluation of a function.  But IIUC there isn't actually a known function for H.  `Show Proof` seems uninformative.

Does that make sense?  Where is a good place to put the new section in the document?  The CIC chapter is much too dense; it would not fit there.

A simpler example:
```
Goal forall x: nat, (forall n, n >= 0) -> True.
intros.  (* H is forall n : nat, n >= 0 *)
Show Proof.
(*
  (fun (x : nat) (H : forall n : nat, n >= 0) =>
 ?Goal@{x:=x; H:=H})
*)
specialize (H x). (* H is x >= 0 *)
Show Proof.
(*
  (fun (x : nat) (H : forall n : nat, n >= 0) =>
   let H0 : x >= 0 := H x in ?Goal@{x:=x; H:=H0})
*)
```

Reply:
Formally defining "type" conflicts with how it's commonly use in describing computer languages (including OCaml).  Formally the type of `(Nat.add 1 1)` is `forall ...`, which conflicts with the saying that the type of `Nat.add` is a function.  We can only say, so we can say `Nat.add` _is_ a function or `(Nat.add 1) _is_ a function.  Maybe I'll clarify that at some point.

The essential point is that "it's time to create a section that starts to explain the relationship between terms and their types", though probably not today.

Reply:
> Formally defining "type" conflicts with how it's commonly use in describing computer languages (including OCaml).

No, it doesn't. This is the same notion, just more powerful types are allowed in Coq as they can refer to terms.

> Formally the type of `(Nat.add 1 1)` is `forall ...`, which conflicts with the saying that the type of `Nat.add` is a function.

That isn't correct. The type of `(Nat.add 1 1)` is `nat`, as reported by `Check`. `Nat.add` is a function (not even a dependent function), it has type `nat -> nat -> nat`.

> We can only say `Nat.add` _is_ a function or `(Nat.add 1)` _is_ a function, dropping back to syntax.

I don't understand what you mean by dropping back to syntax here, but for sure, `Nat.add` is a function and it still is a function when it is partially applied to a single argument.

> The essential point is that "it's time to create a section that starts to explain the relationship between terms and their types", though probably not today.

Actually there is one, but it is unfortunately the most complex one, so I've been reluctant to work on it. Still, it should be modified instead of a new section created. It's the "Typing rules" section (a.k.a., the CIC chapter): https://coq.inria.fr/refman/language/cic.html

Reply:
> Formally the type of `(Nat.add 1 1)` is `forall ...`,

I don't understand.  The `Check (Nat.add 1 1)` says `nat`, not `forall ...` for the type.

> the saying that the type of `Nat.add` is a function

We don't say "the type of `Nat.add` is a function" but instead "the type of `Nat.add` is a function type".  (Types are never functions.  But types can be "function types", i.e., "types whose terms are functions".  Perhaps this is confusing terminology, but I believe it is standard.  (This is standard in colloquial English too.  If you ask about the "racing section" of a car dealership, nobody will think that you're asking for the section that participates in races, but will know that you mean for "racing" to apply to the cars in the section.  Similarly if you ask for the "on sale aisle" of a grocery store.  The aisle itself is not on sale.  Similarly, since types can never *be* functions, we infer that "function" must apply to terms of that type.))

Reply:
> saying that the type of `Nat.add` is a function.

Nobody ever says that. I insist: nodody says that. The TYPE of Nat.add IS NOT a function. On the other hand Nat.add ITSELF is a function.

Reply:
OK, I got the types wrong.  I want to avoid diving deeply into describing these details in this PR, which I think will take time to get right.  The original PR was submitted 14 months ago, parts of which were split off and merged a while ago.

> > The essential point is that "it's time to create a section that starts to explain the relationship between terms and their types", though probably not today.

> Actually there is one, but it is unfortunately the most complex one, so I've been reluctant to work on it. Still, it should be modified instead of a new section created. It's the "Typing rules" section (a.k.a., the CIC chapter): https://coq.inria.fr/refman/language/cic.html

That's what I intended.  (Perhaps I need to think more clearly so you can more easily read my mind :-)  "section" can include "subsection", "subsubsection", ....  As a first step, that chapter needs an introduction that makes sense to the non-expert.  That's going to take a focused effort and is beyond the scope of this PR.

> > saying that the type of `Nat.add` is a function.

> Nobody ever says that. I insist: nobody says that. The TYPE of Nat.add IS NOT a function. On the other hand Nat.add ITSELF is a function.

Consider someone with experience programming in OCaml who is learning Coq on their own and is not familiar with Coq typing.  They may carry over ideas from programming that are incorrect for Coq.  The doc should dispel some of the more likely potential misconceptions.  People do refer to "function types" in OCaml, eg [here](https://www2.lib.uchicago.edu/keith/ocaml-class/data.html#function):

> **Functions**

> In any functional language, one of the most important types of data is the function. Functions can be stored in data structures, passed to functions and returned as the result of functions, and computed dynamically.

Programmers say informal things like "what is the type of `v`?  It's a function".  We need to re-educate people a bit to distinguish between what something _is_ and its type.  As I said "We can only say `Nat.add` _is_ a function or `(Nat.add 1) _is_ a function".

Let's not spend a lot more time on this right now.  I think it's not needed in this PR.

Reply:
> Programmers say informal things like "what is the type of `v`? It's a function".

I think people saying that know perfectly well the abuse here. In ocaml a function is a value and a type cannot be a value so the two worlds are completely separated.

> We need to re-educate people a bit to distinguish between what something _is_ and its type. As I said "We can only say `Nat.add` _is_ a function or `(Nat.add 1) _is_ a function".

Note that saying thet `a -> b` is a "function type" is ok because it means "the type of a function".

> Let's not spend a lot more time on this right now. I think it's not needed in this PR.

OK